### PR TITLE
Add pivot updator for optimism

### PR DESCRIPTION
**Pull request by marcindsobczak**
## Changes

- adding `UnsafePivotUpdator` which is working similar as general `PivotUpdator` except using head block hash (minus 64 blocks) instead of finalized block hash.

In `PivotUpdator` we are using finalized block hash, because it can't be reorganized so is a safe starting point. Optimism is not providing finalized block hash until fully synced, so we need to use unsafe head block hash, which potentially can be reorganized and break the node. For more safety we are using head block minus 64. On Optimism reorganizations aren't common, so it can be a justified risk to significantly improve UX in most cases

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by LukaszRozmej** [comment]
Wouldn't it be safer to get Head - N blocks? (Whatever that N would be)

**Comment by marcindsobczak** [comment]
@LukaszRozmej it would be safer, but then we need to refactor and complicate the logic way more

Right now we are getting block hash and we are asking peers for a block by hash. Then we have block number and that's it.

If we want to have `head - N`, then we need to ask peers again about block `head - N` to get it's hash

**Comment by marcindsobczak** [comment]
Ok after discussion we need to use `head - N`, so working on it

**Review by kamilchodola — APPROVED** [review]

**Review by flcl42 — APPROVED** [review]

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    ISyncConfig syncConfig,
+    IBlockCacheService blockCacheService,
+    IBeaconSyncStrategy beaconSyncStrategy,
+    IDb metadataDb,
+    ILogManager logManager)
+    : PivotUpdator(blockTree, syncModeSelector, syncPeerPool, syncConfig,
+        blockCacheService, beaconSyncStrategy, metadataDb, logManager)
+{
+
+    private const int NumberOfBlocksBehindHeadForSettingPivot = 64;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs:34]: `Reorganization.MaxDepth`?

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PivotUpdator.cs [hunk 2]
```diff
@@ -126,38 +126,29 @@ private async void OnSyncModeChanged(object? sender, SyncModeChangedEventArgs sy
...
+    protected virtual async Task<(Hash256 Hash, long Number)?> TryGetPivotData(CancellationToken cancellationToken)
     {
+        // getting finalized block hash as it is safe, because can't be reorganized
         Hash256? finalizedBlockHash = _beaconSyncStrategy.GetFinalizedHash();
 
         if (finalizedBlockHash is null || finalizedBlockHash == Keccak.Zero)
         {
-            if (_logger.IsInfo && (_maxAttempts - _attemptsLeft) % 10 == 0) _logger.Info($"Waiting for Forkchoice message from Consensus Layer to set fresh pivot block [{_maxAttempts - _attemptsLeft}s]");
-
+            PrintWaitingForMessageFromCl();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PivotUpdator.cs:141]: when you invert this `if` then the override looks quite similar in structure and probably can be mostly shared.

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    return (potentialPivotBlockHash, potentialPivotBlockNumber);
+                }
+            }
+        }
+
+        PrintWaitingForMessageFromCl();
+        return null;
+    }
+
+    private long? TryGetHeadBlockNumberFromBlockCache(Hash256 headBlockHash)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs:66]: more or less same as `TryGetFinalizedBlockNumberFromBlockCache`

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                if (_logger.IsInfo) _logger.Info($"Loaded head block {headBlockHash} from block cache. Head block number: {headBlock.Header.Number}");
+                return headBlock.Header.Number;
+            }
+            if (_logger.IsDebug) _logger.Debug($"Hash of header found in block cache is {headBlock.Header.Hash} when expecting {headBlockHash}");
+        }
+
+        return null;
+    }
+
+    private async Task<long> TryGetHeadBlockNumberFromPeers(Hash256 headBlockHash, CancellationToken cancellationToken)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs:82]: more or less the same as `TryGetFinalizedBlockNumberFromPeers`

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                if (_logger.IsInfo) _logger.Info($"Loaded potential pivot block {potentialPivotBlockNumber} from block cache. Hash: {block.Hash}");
+                return block.Hash;
+            }
+        }
+
+        if (_logger.IsDebug) _logger.Debug("Header of pivot block not found in block cache");
+        return null;
+    }
+
+    private async Task<Hash256?> TryGetPotentialPivotBlockHashFromPeers(long potentialPivotBlockNumber, CancellationToken cancellationToken)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/UnsafePivotUpdator.cs:131]: very similar to other peer methods `TryGetPotentialPivotBlockHashFromPeers`

**Review by LukaszRozmej — COMMENTED** [review]
I think the class hierarchy is too complicated duplicating a lot of stuff, details in comments.
This resulted of making a lot of stuff `protected` which is not great. 
Basically subpar code organization and reuse.

**Pull request merged**